### ISSUE TITLE
Is cito's ALE accurate?

### ISSUE THREAD

**Issue by tripartio** [citoverse/cito#119]
I've recently discovered `cito` as my new favourite neural networks package for R. Thanks for your great work on this.

I'm an interpretable machine learning (IML) researcher, so I'm thrilled to see `cito`'s rich set of interpretability and feature importance features. I am particularly focused on accumulated local effects (ALE); I've written the [`ale` package](https://tripartio.github.io/ale/) dedicated to ALE-based calculations, plots, effect sizes, and statistical inference.

As I explored advanced neural network packages in R, I really appreciate cito's excellent design. My model is model-agnostic and I've found that the two packages are effortlessly compatible, unlike [some other "R"-based neural network packages that have complexities like Python dependencies]( https://github.com/tripartio/ale/issues/11) that cause some rough edges.

However, on examining `cito`'s native implementation of ALE, I am wondering if it is accurate. This issue presents an MRE that shows my concern.

``` r
# Load packages for this MRE
library(ALEPlot)
library(purrr)
library(ale)
#> 
#> Attaching package: 'ale'
#> The following object is masked from 'package:base':
#> 
#>     get
library(cito)
#> 
#> Attaching package: 'cito'
#> The following object is masked from 'package:ale':
#> 
#>     ALE
library(patchwork)

# Train a neural network
nn.fit<- dnn(
  Species ~ .,
  data = iris, 
  loss = "cross-entropy",
  verbose = FALSE
)
#> Registered S3 methods overwritten by 'reformulas':
#>   method       from
#>   head.call    cito
#>   head.formula cito
#>   head.name    cito
```

![](https://i.imgur.com/Cvn89YA.png)

``` r

# Create and plot cito::ALE plots
cito_ale <- cito::ALE(nn.fit)
#> Number of Neighborhoods reduced to 8
#> Number of Neighborhoods reduced to 8
#> Number of Neighborhoods reduced to 8
```

<img width="1064" height="1001" alt="Image" src="https://github.com/user-attachments/assets/b9ec12ec-2da2-4548-816c-dcfb46260150" />

To verify the accuracy of ALE data and plots, I always refer to the reference `ALEPlot` package by the original creator of ALE. Unfortunately, [the package has been delisted from CRAN](https://cran.r-project.org/web/packages/ALEPlot/index.html), but the latest 1.1 version can still be [downloaded and installed from the CRAN archive](https://cran.r-project.org/src/contrib/Archive/ALEPlot/).

``` r

# Plot the reference ALEPlot::ALEPlot plots in the same orientation as cito::ALE
# output, since it only prints individual base R plots.
par(mfrow = c(4, 3))

# Iterate by variables in columns 1 to 4:
# Sepal.Length, Sepal.Width, Petal.Length, Petal.Width
for (i.var in 1:4) {
  
  # Iterate by species:
  # setosa, versicolor, virginica
  for (i.species in 1:3) {
    ALEPlot::ALEPlot(
      X = iris[, -5],
      X.model = nn.fit,
      pred.fun = function(X.model, newdata) {
        predict(X.model, newdata, type = 'response')[, i.species]
      },
      J = i.var,
      K = 10
    )
  }
}
```
<img width="973" height="1055" alt="Image" src="https://github.com/user-attachments/assets/4bb4f098-263b-4b75-9807-f0464dcaf55d" />

Comparing the ALE plots from `cito` and `ALEPlot`, the results for `Sepal.Length`, `Sepal.Width`, and `Petal.Width` look fine; they are essentially identical. However, the results for `Petal.Length` are markedly different:

* The ALE y ranges for all three species are different.
* The shape of `Petal.Length` for versicolor is quite different (this is the only plot with a different shape).

I don't think there is an issue with the original `ALEPlot` implementation; my `ale` package reproduces the original data, as seen in the following code. (Carefully note that even though [the ALE y scale is uniformized from 0 to 1 for all plots](https://github.com/citoverse/cito/issues/37), the range of the values and shapes are identical to those from the `ALEPlot` plots above.)

``` r

# Create an ale::ALE object from the cito::dnn model object.
# ale package defaults work perfectly with cito
ale_ale <- ale::ALE(nn.fit)

# Plot the ale::ALE object in the same orientation as cito::ALE output,
# since its default is oriented differently.
ale_ale |> 
  plot(ale_centre = 'zero') |>  # Create ale::ALEPlots object
  get() |>  # Convert to list of ggplots
  # Change orientation to match cito::ALE
  map(~ .x$d1) |> 
  list_transpose() |> 
  flatten() |> 
  patchwork::wrap_plots(ncol = 3)
#> Warning in annotate(geom = "label", x = y_summary["max"], y = which(estimates$aler_max == : Ignoring unknown parameters: `label.size`
#> Ignoring unknown parameters: `label.size`
#> Ignoring unknown parameters: `label.size`
```

<img width="973" height="1055" alt="Image" src="https://github.com/user-attachments/assets/360c368a-227f-48ab-b716-29cbd1b6e5bc" />

<sup>Created on 2025-10-20 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

I'm not sure what the source of the disparity would be. One thought is perhaps the `cito` ALE results are not based on the full training dataset. But I don't know the details of its implementation.

**Comment by MaximilianPi** [comment]
Hi @tripartio,

Thank you for your feedback! 
Your package is interesting, I will explore it! (Did you come up with a quicker way to get standard errors than bootstrapping?)

It appears that our different default options for partitioning the feature space are the cause. We support equal distances (_ALE_type = "equidistant",_ our default) and quantiles (`ALE_type = "quantile"`). If you switch to "quantile," you will get the expected shapes:

`ALE(nn.fit, type = "response", variable = "Petal.Length", ALE_type = "quantile")`.



<img width="1189" height="537" alt="Image" src="https://github.com/user-attachments/assets/fded5fbc-37fb-4a82-82c5-e851f1bf2d74" />

However, I am not sure why equidistance is our default. It enforces reweighting in favor of sparse regions of the feature space. I'll have to think about that... At the very least, we should improve the documentation on this topic. 

(But there seems to be small bug because I cannot run `ALE(nn.fit,type = "response", ALE_type = "quantile")`, nevertheless it works for specific variables. )

**Comment by tripartio** [comment]
Wow, thanks for the super-fast response. 

I've never considered equidistance bins; as you noted, that would misleadingly exaggerate sparse spaces. It nullifies what rug plots try to do: give the analyst a more accurate indication of which ALE values and regions are actually supported by the data.

I suggest that you switch to quantiles as the default. 

In any case, I'm glad that the issue was so easily resolved! Thanks again for the great package. Let me know if you have any questions about mine. I'm glad to see that the two packages work so well together.
